### DISCUSSION TITLE
Node Server GModelFactory

### DISCUSSION THREAD

**Discussion by csamarchi** [eclipse-glsp/glsp#721] [General]
Hi there,

I have a Node server VsCode extension graphical model going. I am using a YAML file and parsing it with our own parsing method and it is working, I am able to parse our file, loop through, and build instances of our model types and add them to the model state. See the function below: 

![Screen Shot 2022-07-27 at 5 25 36 PM](https://user-images.githubusercontent.com/42501494/181383112-da6e93d3-885c-4af8-9954-2f94af2742b0.png)

I'm trying to figure out the next steps and wondering if I'm missing something. Is there another step needed on the server side or is the rest done in the client? I know Node projects do not use GModelFactory, they must be custom, and I think that might be the the next step before rendering/building out on the client side, but wanted to get a better explanation?

Our GModelFactory is following the format of the TaskList example [here](https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/node-json-vscode/glsp-server/src/model/tasklist-gmodel-factory.ts).

Here is the erroring I'm running into (thought it could provide some extra context):
![Screen Shot 2022-07-27 at 5 24 17 PM](https://user-images.githubusercontent.com/42501494/181382920-46cbb560-ee18-4004-9e95-88592bf75424.png)

Any help is much appreciated

**Comment by planger** [comment]
Hi @csamarchi,

after you have successfully integrated your [custom source model](https://www.eclipse.org/glsp/documentation/sourcemodel/), the next step is to [generate a GModel with a custom `GModelFactory`](https://www.eclipse.org/glsp/documentation/gmodel/) from it.

It looks like your current `GModelFactory` doesn't set a proper GModel, which is why the error is thrown when the server tries to send it to the client. Once your custom `GModelFactory` has created a GModel from your YAML files, it needs to set it via `modelState.updateRoot(newModel);`.

The client doesn't know your YAML files and can only process and render GModels, which are created on the server from your YAML files.

See also https://www.eclipse.org/glsp/documentation/gmodel/

Hope this helps!

**Reply by csamarchi** [reply]
So, I first need to set a GModel and then send it to the `modelState.updateRoot()` method and then the client takes over from there? 

Curious, how you were able to tell from the error that the GModel was not created properly? What is wrong with the below line? 
`  const moduleNodes = modules.map(module => this.createModule(module));
`

**Reply by planger** [reply]
Right, you need to generate the GModel from your model source and write it into the model state via `modelState.updateRoot()`. The submission handler will then take it from there and submit it.

Well I just saw that the model root is undefined in the submission handler, which indicates that it has never been set in the model state. Of course this is just my best guess. :-)

**Reply by planger** [reply]
See also in the project template how the factory eventually sets the generated GModel to the model root: https://github.com/eclipse-glsp/glsp-examples/blob/00185505c83bef731c5a9435d01ced3d896d95f2/project-templates/node-json-theia/glsp-server/src/model/tasklist-gmodel-factory.ts#L31

**Reply by csamarchi** [reply]
Thanks for all the help, still wondering how you got model root from the error, `Cannot read property 'revision' of undefined` and how I'm not setting my GModel correctly...


```
 createModel(): void {
        const modules = this.modelState.modules;
        this.modelState.index.indexModule(modules);
        const moduleNodes = modules.map(module => this.createModule(module));
        const newRoot = GGraph.builder().id(module.id).addChildren(moduleNodes).build();
        this.modelState.updateRoot(newRoot);
    }
```

**Reply by planger** [reply]
Yeah, I remembered that [`revision`](https://github.com/eclipse-glsp/glsp-server-node/blob/3996af680ef2904f6b7ac30ab7d21145d95f55d7/packages/graph/src/gmodel-element.ts#L198) is a property of `GModelRoot`. If now `revision` can't be read from `undefined`, that indicates that the model root hasn't been set. But of course, I may certainly be wrong.

**Reply by csamarchi** [reply]
I'm so confused why though. What is wrong with how I'm setting my model root? I'm getting no other errors other than can't read `revision` of `undefined`. I'm following the docs exactly and if what the doc says I should be setting up my GModelFactory and model root properly.

**Reply by tortmayr** [reply]
Hi @csamarchi,
your implementation looks correct to me. As far as I can see you implemented everything correctly and according to the docs. What version of the glsp-server-node are you using? I had a look at the recent changes and I fear that we might have introduced a regression bug with a [recent change](https://github.com/eclipse-glsp/glsp-server-node/commit/afa6928bf96daae2a28e5267a635c25cb393db53#r79685044).
We have to look into this but for the moment you might be able to work around this issue with manually initializing the revision with `0´:
```ts
        const newRoot = GGraph.builder().id(module.id).addChildren(moduleNodes).revision(0).build();
```

**Reply by csamarchi** [reply]
`"@eclipse-glsp/server-node": "0.9.0-next.b178eb2.24"`

**Reply by tortmayr** [reply]
The GLSP documentation is based on the latest release (1.0.0) so I would recommend to update to the release version (if possible):
```
"@eclipse-glsp/server-node": "1.0.0"
```

**Reply by csamarchi** [reply]
I updated to the latest release and used that workaround and still getting same error

**Reply by csamarchi** [reply]
Any updates to the regression bug mentioned above?

**Reply by tortmayr** [reply]
Since you are using version 1.1.0 the regression bug mentioned above should not affect you.

**Comment by csamarchi** [comment]
Also - we are using the Workflow example and TaskList example for our node server vscode extension scenario. I'm aware that the workflow example does not use a GModelFactory instead uses the so-called NullImpl of the GModelFactory. Does that have any affect on our use case?

**Reply by planger** [reply]
Not really, I think. But in the workflow example, the used `SourceModelStorage` [already writes the source model directly into the model state's root](https://github.com/eclipse-glsp/glsp-server-node/blob/3996af680ef2904f6b7ac30ab7d21145d95f55d7/packages/server-node/src/gmodel-lib/gmodel-storage.ts#L45), So the `NullImpl` of the `GModelFactory` doesn't need to do that anymore and still everything works.
If you now however replace the `SourceModelStorage` with an own implementation that "just" writes the source model into the model state, there is still no `GModel` in it. Therefore, your `GModelFactory` needs to generate and put it there.

**Comment by rsoika** [comment]
@csamarchi I have written a short tutorial about the concept of the GmodelFactory and SourceModelStorage. We use a java implementation to load and build our own model. Maybe this can be helpful for your understanding:

https://ralph.blog.imixs.com/2022/05/20/build-your-own-modelling-tool-with-eclipse-glsp/
